PhabricatorBMOAuthProvider.php

```diff
@@ -0,0 +1,418 @@
+    $json_obj = array('result' => $trans_code);
+
+    // Render the response JSON
+    $response = id(new AphrontJSONResponse())
+      ->setAddJSONShield(false)
+      ->setHTTPResponseCode(200)
+      ->setContent($json_obj);
+    return array($account, $response);
+  }
+
+  private function processLoginRequestConfirmationGet($controller, $request) {
+    $response = null;
+    $csrf = $controller->getExtraURIData();
+
+    // Verify with CSRF as an additional security measure
+    $this->verifyAuthCSRFCode($request, $csrf);
+
+    // Match result token and client_api_login to find client_api_key
+    $provided_trans_code = $request->getStr('callback_result');
+    $provided_api_login = $request->getStr('client_api_login');
+
+    $token = id(new PhabricatorAuthTemporaryTokenQuery())
+      ->setViewer(PhabricatorUser::getOmnipotentUser())
+      ->withTokenResources(array($provided_trans_code))
+      ->withTokenCodes(array($csrf))
+      ->withTokenTypes(array(self::TOKEN_TYPE))
+      ->withExpired(false)
+      ->executeOne();
+
+    // No token means we've received invalid information from Bugzilla
+    if(!$token) {
+      $this->throwException(pht('No temporary token found for this transaction code (%s) or CSRF token (%s).', $provided_trans_code, $csrf));
+    }
+
+    // Compare result token and client_api_login to original response
+    $token_props = $token->getProperties();
+    if($token_props['client_login'] != $provided_api_login) {
+      $this->throwException(
+        pht('Token\'s API Login does not match Bugzilla API Login.')
+      );
+    }
+
+    // Call BMO Who Am I REST resource to validate API Key (X-Bugzilla-API-Key)
+    $adapter = $this->getAdapter();
+    $config = $this->getConfig();
+    $api_key = $token_props['api_key'];
+
+    $future_uri = id(new PhutilURI(''))
+      ->setProtocol(
+        $config->getProperty(self::CONFIG_KEY_BUGZILLA_PROTOCOL)
+      )
+      ->setDomain($config->getProperty(self::CONFIG_KEY_BUGZILLA_DOMAIN))
+      ->setPath('/rest/whoami');
+
+    $future = id(new HTTPSFuture((string) $future_uri))
+      ->setMethod('GET')
+      ->addHeader('X-Bugzilla-API-Key', $api_key)
+      ->addHeader('Accept', 'application/json')
+      ->setTimeout(5);
+
+    // Resolve the async HTTPSFuture request and extract JSON body
+    $whoami_body = '';
+    try {
+      list($whoami_body) = $future->resolvex();
+    } catch (HTTPFutureResponseStatus $ex) {
+      $this->throwException(pht('Bugzilla WhoAmI request failed to resolve.'));
+    }
+
+    $user_json = array();
+    try {
+      $user_json = phutil_json_decode($whoami_body);
+    }
+    catch(Exception $e) {
+      $this->throwException(
+        pht('JSON from Bugzilla WhoAmI could not be parsed: '.$whoami_body)
+      );
+    }
+
+    // If there's no "id" key in the JSON, we know something is wrong
+    if(!isset($user_json['id'])) {
+      $this->throwException(pht('No user ID was provided by Bugzilla.'));
+    }
+
```

Comment:
Assuming this is the end of the login process, can we log an information message?  Something like "Successful BMO Login for user X".